presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java

```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
         }
     }
 
-    @ThreadSafe
+    /**
+     * Each LookupSource has it's own copy of OuterPositionTracker instance.
+     * Each of those OuterPositionTracker must be committed after last write
+     * and before first read.
+     *
+     * All instances share visitedPositions array, but it is safe because each thread
+     * starts with visitedPositions filled with false values and marks only some positions
+     * to true. Since we don't care what will be the order of those writes to
+     * visitedPositions, writes can be without synchronization.
+     *
+     * Synchronization in commit() is required because each writing thread must have
```

Comment:
>  I would have to add comment to referenceCount variable: "DO NOT REMOVE IT, because it's synchronization is required to guarantee memory visibility between commit and getVisitedPositions"

You'd have to do that for `synchronized` on the method, regardless. Otherwise, an empty synchronized method would look odd and someone might end up removing it.

I think the reference count is a smart way of encoding the required semantics:
* If `getVisitedPositions` saw all the decrements, then it must see the updates to the positions.
* If `getVisitedPositions` didn't see all the decrements, there's a sequencing bug and the method will fail loudly.